src/cliffordsynthesis/encoding/MultiGateEncoder.cpp

```diff
@@ -132,19 +134,61 @@ void MultiGateEncoder::assertTwoQubitGateOrderConstraints(
   }
 
   // gate variables of the current and the next time step
-  const auto& current = vars.gC[pos][ctrl][trgt];
-  const auto& gSNow   = vars.gS[pos];
-  const auto& gCNext  = vars.gC[pos + 1];
+  const auto& gSNow  = vars.gS[pos];
+  const auto& gCNext = vars.gC[pos + 1];
 
   // two identical CNOTs may not be applied in a row because they would cancel.
-  lb->assertFormula(LogicTerm::implies(current, !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][ctrl][trgt], !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][trgt][ctrl], !gCNext[trgt][ctrl]));
+
+  // no gate on both qubits => no CNOT on them in the next time step.
+  constexpr auto noneIndex = gateToIndex(qc::OpType::None);
+  const auto     noGate    = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
+  // similarly,
+  // H - X   c - H
+  //     | = |
+  // H - c   X - H
+  // i.e., Hadamards on both qubits => no CNOT on them in the next time step.
+  constexpr auto hIndex = gateToIndex(qc::OpType::H);
+  const auto     hh     = gSNow[hIndex][ctrl] && gSNow[hIndex][trgt];
 
-  // if no gate is applied to both qubits, no CNOT on them can be applied in the
-  // next time step.
-  const auto noneIndex     = gateToIndex(qc::OpType::None);
-  const auto noGate        = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
   const auto noFurtherCnot = !gCNext[ctrl][trgt] && !gCNext[trgt][ctrl];
-  lb->assertFormula(LogicTerm::implies(noGate, noFurtherCnot));
+  lb->assertFormula(LogicTerm::implies(noGate || hh, noFurtherCnot));
+
+  // No Combination of Paulis or Hadamards on both Qubits before a CNOT
+  // These gates can be just pushed through to the other side
+
+  if (pos == 0) { // In the beginning parallel gates are allowed
+    return;
+  }
```

Comment:
I am fairly sure that this condition is not necessary. Why shouldn't these rules be allowed in the very beginning. The scenario in the beginning is just the same as at any other point in the circuit.
A block of two single-qubit gates (paulis+H) and a CNOT can _always_ be replaced by a CNOT and at most two single-qubit gates.
I could image what you were seeing was a consequence of the duplicate rules highlighted in the other comment.

Reply:
Only parallel H and Paulis are a problem so I added a special case for the first timestep that encodes only the other combinations

Reply:
No but you could push the Pauli from the end to the beginning to reduce the depth by 1. But the symmetry breaker wouldn't allow that. That's why we have to exclude the first layer.